### ISSUE TITLE
Mention vim.api.nvim_call_function

### ISSUE THREAD

**Issue by clason** [nanotee/nvim-lua-guide#6]
especially the difference to `vim.fn`/`vim.call`:
```
vim.fn.{func}({...})                                    vim.fn
        Invokes vim-function or user-function {func} with arguments {...}.
        To call autoload functions, use the syntax:
            vim.fn['some#function']({...})

        Unlike vim.api.nvim_call_function this converts directly between Vim
        objects and Lua objects. If the Vim function returns a float, it will
        be represented directly as a Lua number. Empty lists and dictionaries
        both are represented by an empty table.

        Note: v:null values as part of the return value is represented as
        vim.NIL special value

        Note: vim.fn keys are generated lazily, thus pairs(vim.fn) only
        enumerates functions that were called at least once.
```
so I _think_ you'd always want to use `vim.call`?

**Comment by matu3ba** [comment]
`:help vi.call`

```txt
vim.call({func}, {...})					*vim.call()*
        Invokes |vim-function| or |user-function| {func} with arguments {...}.
        See also |vim.fn|.
        Equivalent to: 
            vim.fn[func]({...})
```

It is the same. Closing?

**Comment by matu3ba** [comment]
@nanotee If possible, one shall always use vim.cmd or vim.api (because they dont need to be parsed or something). However, you cant call user functions with that.

**Comment by nanotee** [comment]
From what I gathered this is really context dependent, and even then I'd be surprised if the difference was noticeable in most cases. I don't know if this is all that useful to mention in a "getting started" guide

**Comment by clason** [comment]
I forgot that I originally opened this issue 🤦 

The difference between `vim.call` and `vim.fn` (the latter is just syntactic sugar on top of the former for a more "Lua look") is already well explained in the guide; there's nothing that should be added or changed there.

It _might_ be useful to mention that if the function is part of the nvim API (i.e., available through `vim.api.*`), calling that directly rather than through `vim.fn.*` is in general preferable as this avoids a roundtrip. (I don't think it will be a noticeable performance difference, but the additional wrapping _may_ make a difference for things like scheduling.)

`vim.cmd` is another kettle of fish entirely and a red herring in this context.

**Comment by matu3ba** [comment]
@clason 

> It might be useful to mention that if the function is part of the nvim API (i.e., available through vim.api.*), calling that directly rather than through vim.fn.* is in general preferable as this avoids a roundtrip. (I don't think it will be a noticeable performance difference, but the additional wrapping may make a difference for things like scheduling.)

This belongs to the advanced guide, as it is a performance optimisation in hot code paths. I am not sure either, if it makes a difference in jitted code imho.

> `vim.cmd` is another kettle of fish entirely and a red herring in this context.

I dont understand. `vim.api.xxx` is most performant, as the other wrap and parse the types. The help page is quite unprecise on what is happening though. So this belongs to the advanced guide in #43 .

**Comment by clason** [comment]
> This belongs to the advanced guide, as it is a performance optimisation in hot code paths. I am not sure either, if it makes a difference in jitted code imho.

I disagree; it's important to know how the different methods relate (e.g., that `vim.fn` and `vim.call` are exactly the same but `vim.api` is different under the hood). And as I wrote, the point is not performance but actual behavior (i.e., some things work with one but not the other). You don't need to go into details and make a table or so, but it's something I think beginners should be aware of.

> > `vim.cmd` is another kettle of fish entirely and a red herring in this context.
> 
> I dont understand. `vim.api.xxx` is most performant, as the other wrap and parse the types. The help page is quite unprecise on what is happening though. So this belongs to the advanced guide in #43 .

The point is that `vim.cmd` allows executing _arbitrary_  vimscript (as vimscript), while `vim.api` only allows calling neovim API functions from Lua code. Very different use cases -- there are things you _can't_ do with `vim.api` (such as setting autocommands).

**Comment by matu3ba** [comment]
@clason Thats true. You can not use internal functions from `vim.api`. Do you have a suggestion text for [calling-vimscript-functions](https://github.com/nanotee/nvim-lua-guide#calling-vimscript-functions) ?

**Comment by clason** [comment]
I'd suggest to add the following short notes to https://github.com/nanotee/nvim-lua-guide#vimfnfunction :

1. Before the link to the help, add a sentence that `vim.fn` is functionally identical to `vim.call` but allows a more Lua-like syntax.
2. In `Tips`, add a second bullet point that API functions from `:h api` are also available directly through `vim.api`; e.g., `vim.fn.nvim_list_uis()` and `vim.api.nvim_list_uis()` both work, and that in general the latter is preferable.

**Comment by matu3ba** [comment]
After short discussion with TJ Devries (one core dev of neovim), we removed 1. in 16423fe3a985473da782134f05940a4dd527241e as it might get deprecated.

**Comment by clason** [comment]
> @clason After short discussion with TJ Devries (one core dev of neovim), we removed 1. in [16423fe](https://github.com/nanotee/nvim-lua-guide/commit/16423fe3a985473da782134f05940a4dd527241e) as it might get deprecated.

No, that's not going to happen; `vim.fn` is implemented on top of `vim.call` -- if the latter goes away, so does the former. The point is that people should be discouraged from using `vim.call` in favor of `vim.fn`, and this is exactly something that should be in this guide.

Also, `vim.api.X` may in fact not be the optimal way **if** `vim.X` also exists. So the rule of thumb should be

`vim.X` >> `vim.api.X` >> `vim.fn.X`

(assuming these exist; i.e., you'd only use a later one if the previous one doesn't exist).

**Comment by cloggier** [comment]
@clason thats very concise and informative way to put it. It should be added to the guide I think.

**Comment by matu3ba** [comment]
@clason Good enough like this? I would prefer to explain internal functions into the advanced guide, ie for plugin creation.

**Comment by clason** [comment]
Yes, that's good.

**Comment by nanotee** [comment]
I have finally merged #49, apologies for the delay.

Looking at the backtrace for `nvim_list_uis()` with `vim.fn` vs `vim.api` in gdb, it does seem like the `vim.fn` version is doing some extra work, so that's good to know.

I believe the original issue was about `nvim_call_function()` vs `vim.call/vim.fn` specifically, which is a different matter entirely. I have yet to encounter a situation where I'd use `nvim_call_function()` over `vim.fn`, so I don't know if I should mention it here

**Comment by clason** [comment]
from `:h vim.fn` (which just calls `vim.call`):

>  Unlike vim.api.nvim_call_function this converts directly between Vim
        objects and Lua objects. If the Vim function returns a float, it will
        be represented directly as a Lua number. Empty lists and dictionaries
        both are represented by an empty table.

So I don't know where you'd use that, either -- maybe if you want to do some sort of transformation of vimscript code? That wouldn't be anything for a beginner's guide, though 😆

**Comment by matu3ba** [comment]
> I believe the original issue was about `nvim_call_function()` vs `vim.call/vim.fn` specifically, which is a different matter entirely. I have yet to encounter a situation where I'd use `nvim_call_function()` over `vim.fn`, so I don't know if I should mention it here

It would be great, if you could examplify "what extra work" you mean or close the issue.

**Comment by nanotee** [comment]
> It would be great, if you could examplify "what extra work" you mean or close the issue.

In my tests this resulted in 2 extra function calls for `vim.fn` vs using `vim.api` directly when printing the backtrace, but now that using API functions via `vim.fn` is no longer allowed, this is kinda moot

I have addressed `nvim_call_function` in #56, hopefully it's sufficiently detailed. Thanks everyone for your input, I genuinely appreciate the help!
